<a href="https://colab.research.google.com/github/eormeno12/CursoIntroduccionAlML/blob/6.Collab_con_Scikit/6_Collab_con_Scikit.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Collab con Scikit

---

Clasificación de flores.

Set de datos de Scikit -> https://scikit-learn.org/stable/datasets/index.html

In [3]:
#importa datasets de sklearn
from sklearn.datasets import load_iris
#importa el algoritmo RandomForestClassifier (C L A S I F I C A D O R)
from sklearn.ensemble import RandomForestClassifier
#maneja datos
import pandas as pd
#maneja números
import numpy as np

Lo que hace **random.seed** es que genera un pre set de datos para que si compartimos el notebook las otras personas puedan tener resultados similares, evitamos incertidumbres.

In [4]:
np.random.seed(0)

In [5]:
iris = load_iris()

In [7]:
dataframe = pd.DataFrame(iris.data, columns=iris.feature_names)

In [8]:
dataframe.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

Preparación de datos

In [9]:
dataframe["species"] = pd.Categorical.from_codes(iris.target, iris.target_names)

In [11]:
dataframe.head()

sepal length (cm)  sepal width (cm)  ...  petal width (cm)  species
0                5.1               3.5  ...               0.2   setosa
1                4.9               3.0  ...               0.2   setosa
2                4.7               3.2  ...               0.2   setosa
3                4.6               3.1  ...               0.2   setosa
4                5.0               3.6  ...               0.2   setosa

[5 rows x 5 columns]

**80%** datos de entrenamiento y **20%** de prueba.

In [12]:
train_random = np.random.rand(len(dataframe)) < 0.8
train_set = dataframe[train_random]

La ~ hace que nos entregue el otro porcentaje de datos.

In [14]:
test_set = dataframe[~train_random]

In [15]:
print(f"Number obs en el set de entrenamiento = {len(train_set)}")

Number obs en el set de entrenamiento = 123


In [17]:
print(f"Number obs en el set de prueba = {len(test_set)}")

Number obs en el set de prueba = 27


In [18]:
features = dataframe.columns[:4]
features

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

El **factorize** nos permite asignarle un valor númerico a la variable string de nuestro target.

In [23]:
y = pd.factorize(train_set["species"])[0]
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

###Creando el algoritmo

clf -> nombre de convención para un clasificador.

In [25]:
clf = RandomForestClassifier(n_jobs=2, random_state=0)

fit -> función de Scikit para entrenar el modelo.

Se le pasa como primera variable los datos y como segunda las etiquetas si se trata de un algortimo supervisado.

In [26]:
clf.fit(train_set[features], y)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=2,
                       oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

Ahora predecimos.

In [27]:
clf.predict(test_set[features])

array([0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 2, 2, 1, 1, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2])

Convertimos los valores númericos en texto de acuerdo al data set.

In [33]:
preds = iris.target_names[clf.predict(test_set[features])]
preds[:5]

array(['setosa', 'setosa', 'setosa', 'setosa', 'setosa'], dtype='<U10')

In [31]:
test_set["species"].head()

7     setosa
8     setosa
13    setosa
17    setosa
19    setosa
Name: species, dtype: category
Categories (3, object): ['setosa', 'versicolor', 'virginica']

#Matriz de Confusión

Es una herramienta para validar datos en la cual se posicionan los datos reales en un eje mientras que los predichos en otro. De esta forma, si los datos coinciden en un punto, se puede saber si acertaron o no.

Se puede apreciar como se forma una diagonal.

En este caso nuestro modelo falló al identificar a 2 versicolor como virginica.

In [35]:
pd.crosstab(test_set["species"], preds, rownames=["Actual Species"], colnames=["Predicted Species"])

Predicted Species  setosa  versicolor  virginica
Actual Species                                  
setosa                  8           0          0
versicolor              0           5          2
virginica               0           0         12

Ver a cual de los features el clasificador le dio más importancia o peso.

In [38]:
list(zip(train_set[features], clf.feature_importances_))

[('sepal length (cm)', 0.0889479467793623),
 ('sepal width (cm)', 0.0234427551884027),
 ('petal length (cm)', 0.44105878808840404),
 ('petal width (cm)', 0.44655050994383105)]

Esto es como un análisi post-muerte.

**Por ejemplo:**

No podríamos omitir en una investigación el width del petalo porque este tiene un peso de 44%, es sumamente importante para determinar a que tipo de flor de refiere, mientras que con el width del sepalo que solo tiene un 2% de importancia, nos podriamos dar el lujo de excluirlo y saber que el resultado no se va a ver afectado en gran magnitud.